### PR TITLE
[JN-362] [JN-365] Add admin UI for requesting kits and initial DSM integration

### PR DESCRIPTION
**Pull request by breilly2**
Definitely some rough edges here that will be smoothed out with future tickets. Notable changes:
* Kit type entity instead of string value. This allows us to describe them better and to configure studies with them. These are global config since the kit types that Juniper can use are tied to what GP is set up to handle.
* Initial UI for requesting kits and viewing kit requests
* Initial integration with DSM

### CODE REVIEW

File: core/build.gradle [hunk 1]
```diff
@@ -23,6 +23,7 @@ dependencies {
     implementation group: 'org.jdbi', name: 'jdbi3-sqlobject', version: '3.34.0'
     implementation group: 'org.jdbi', name: 'jdbi3-postgres', version: '3.34.0'
     implementation 'org.springframework.boot:spring-boot-starter-data-jdbc:2.7.10'
+    implementation 'org.springframework.boot:spring-boot-starter-webflux:2.7.10'
```
Comment (breilly2) [inline core/build.gradle:26]: This is for Spring's `WebClient`, which may be a bit more fully featured than what we need, but since we're using Spring already, this is what we get out-of-the-box.

File: populate/src/main/java/bio/terra/pearl/populate/service/KitRequestPopulator.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import bio.terra.pearl.core.dao.kit.KitTypeDao;
+import bio.terra.pearl.core.model.kit.KitRequest;
+import bio.terra.pearl.populate.dto.kit.KitRequestDto;
+import bio.terra.pearl.populate.service.contexts.FilePopulateContext;
+import org.springframework.stereotype.Service;
+
+import java.io.IOException;
+import java.util.Optional;
+
+// TODO: finish implementing this
```
Comment (breilly2) [inline populate/src/main/java/bio/terra/pearl/populate/service/KitRequestPopulator.java]: I've been winging it with these populators. They all seem a little bit different and the semantics of what each of the method implementions should do isn't 100% clear to me. Let me know if `KitTypePopulator` looks right and I'll come back and fill this in.

File: populate/src/main/java/bio/terra/pearl/populate/service/StudyPopulator.java [hunk 3]
```diff
@@ -60,6 +67,9 @@ public StudyPopulator(StudyService studyService,
         this.preEnrollmentResponseDao = preEnrollmentResponseDao;
         this.portalDiffService = portalDiffService;
         this.studyUpdateService = studyUpdateService;
+        this.portalEnvironmentService = portalEnvironmentService;
```
Comment (breilly2) [inline populate/src/main/java/bio/terra/pearl/populate/service/StudyPopulator.java:70]: Drive-by bug fix: This wasn't previously being initialized.

File: ui-admin/src/study/participants/KitRequests.tsx [hunk 4]
```diff
@@ -0,0 +1,71 @@
+import React, { useMemo, useState } from 'react'
+import { Enrollee, KitRequest } from 'api/api'
+import { StudyEnvContextT } from '../StudyEnvironmentRouter'
+import { ColumnDef, getCoreRowModel, useReactTable } from '@tanstack/react-table'
+import { basicTableLayout } from 'util/tableUtils'
+import RequestKitModal from './RequestKitModal'
+import { faPlus } from '@fortawesome/free-solid-svg-icons'
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
+import {instantToDefaultString} from "../../util/timeUtils";
```
Comment (devonbush) [inline ui-admin/src/study/participants/KitRequests.tsx]: change to absolute import and single quotes
```suggestion
import {instantToDefaultString} from 'util/timeUtils';
```

File: ui-admin/src/api/api.tsx [hunk 5]
```diff
@@ -435,8 +452,31 @@ export default {
...
+  async createKitRequest(portalShortcode: string, studyShortcode: string, envName: string,
+                         enrolleeShortcode: string, kitType: string): Promise<string> {
+    const params = new URLSearchParams({ kitType })
+    const url =
+      `${baseStudyEnvUrl(portalShortcode, studyShortcode, envName)}/enrollees/${enrolleeShortcode}/requestKit?${params}`
+    const response = await fetch(url, { method: 'POST', headers: this.getInitHeaders() })
+    return await this.processJsonResponse(response)
+  },
+
+  async fetchKitRequests(portalShortcode: string, studyShortcode: string, envName: string,
```
Comment (devonbush) [inline ui-admin/src/api/api.tsx]: maybe rename to `fetchEnrolleeKitRequests` for clarity since we'll likely have a few of these (fetch all, fetch by study, etc..)?

File: ui-admin/src/study/participants/KitRequests.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+                                        onUpdate: () => void
+                                      }) {
+  const [showRequestKitModal, setShowRequestKitModal] = useState(false)
+
+  const onSubmit = () => {
+    setShowRequestKitModal(false)
+    onUpdate()
+  }
+
+  const columns = useMemo<ColumnDef<KitRequest, string>[]>(() => [{
```
Comment (devonbush) [inline ui-admin/src/study/participants/KitRequests.tsx]: declaring these columns outside the component function will save a useMemo call and make it clearer that the column defs don't change in response to the returned data

File: ui-admin/src/study/participants/KitRequests.tsx [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    header: 'Created',
+    accessorKey: 'createdAt',
+    accessorFn: data => instantToDefaultString(data.createdAt)
+  }, {
+    header: 'Address',
+    cell: ({ row }) => <KitRequestAddress sentToAddressJson={row.original.sentToAddress}/>
+  }, {
+    header: 'DSM Status',
+    accessorKey: 'dsmStatus'
+  }], [enrollee.kitRequests])
```
Comment (devonbush) [inline ui-admin/src/study/participants/KitRequests.tsx]: yeah, this isn't the right dependency here -- nothing about the column definitions change with repsect to new data.  So this should be `[]` or, better yet, the declaration pulled outside the component

File: core/src/main/java/bio/terra/pearl/core/config/DataStudyManagerConfig.java [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package bio.terra.pearl.core.config;
+
+import lombok.Getter;
+import lombok.Setter;
+import org.springframework.core.env.Environment;
+import org.springframework.stereotype.Component;
+
+@Component
+@Getter @Setter
+public class DataStudyManagerConfig {
```
Comment (devonbush) [inline core/src/main/java/bio/terra/pearl/core/config/DataStudyManagerConfig.java]: This is just a wrapper around env variables, and isn't (and probably shouldn't) be used anywhere outside of DataStudyManagerClient.  Omitting this class, and moving this to inside the DSM client (or maybe a protected/private helper class inside DSM client, depending on what makes testing easier).  In any case, it definitely belongs in the service/kit directory for now, rather than the general config directory

File: core/src/main/java/bio/terra/pearl/core/service/kit/DataStudyManagerClient.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.algorithms.Algorithm;
+import org.springframework.stereotype.Component;
+import org.springframework.web.reactive.function.client.WebClient;
+import reactor.core.publisher.Mono;
+
+import java.time.Instant;
+
+@Component
+public class DataStudyManagerClient {
```
Comment (devonbush) [inline core/src/main/java/bio/terra/pearl/core/service/kit/DataStudyManagerClient.java]: Let's rename this and the similar files to `PepperDSM______`      "DataStudyManager", even though it's the 'proper' name of the thing, is confusing to read since those are three very general-purpose words strung together. When I first saw "DataStudyManagerConfig" I thought "whoa, we're updating how all study data is managed?"

File: core/src/main/java/bio/terra/pearl/core/service/kit/DataStudyManagerKitRequest.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
+package bio.terra.pearl.core.service.kit;
+
+import lombok.Builder;
+import lombok.Getter;
+
+@Getter @Builder
+public class DataStudyManagerKitRequest {
```
Comment (devonbush) [inline core/src/main/java/bio/terra/pearl/core/service/kit/DataStudyManagerKitRequest.java]: since this class isn't persisted directly, could we just put a PepperKitAddress as a member and save all the field copying?

File: core/src/main/java/bio/terra/pearl/core/model/kit/KitRequest.java [hunk 11]
```diff
@@ -6,14 +6,16 @@
...
 @NoArgsConstructor
 public class KitRequest extends BaseEntity {
     private UUID creatingAdminUserId;
     private UUID enrolleeId;
-    private String kitType;
+    private UUID kitTypeId;
+    private KitType kitType;
     /**
      * JSON blob of address data sent to DSM, collected from Profile/MailingAddress.
      * TODO: decide if this should be separate fields, or maybe use the postgres jsonb type
```
Comment (devonbush) [inline core/src/main/java/bio/terra/pearl/core/model/kit/KitRequest.java]: was the change before, but this TODO should either reference a ticket, or be reworded to "we might"

**Review by devonbush — CHANGES_REQUESTED** [review]
Generally solid CRUD stuff.  Some suggestions for cleanups, the ones I feel most strongly about is calling things "PepperDSM____" rather than "DataStudyManager__"

BLOCKERS:  
1.  In accordance with the new team policy, this needs frontend tests -- don't worry about extensive tests, this more about "at least one reasonable one."  You can check out my [admin email PR](https://github.com/broadinstitute/pearl/pull/400) for an example of how to do a basic "confirm if endpoint is mocked, table appears" test
2.  This needs a backend test that we can successfully link studies to kit types to kits to enrollees.  (essentially confirming that all the DAOs and DB constraints match up).  The test flow would be something like:
a.  create kit type
b. create study
c. assign kit type to study
d. list study kit types
e. create enrollee in study
f. create a kit request for them of the given type (these latter two steps might not be necessary since they're essentially being what's already done in testCreatSampleKit, but they will be necessary if we're going to add server-side logic confirm the kit type for an enrollee is allowed by the study)

**Review by devonbush — CHANGES_REQUESTED** [review]
Generally solid CRUD stuff.  Some suggestions for cleanups, the ones I feel most strongly about is calling things "PepperDSM____" rather than "DataStudyManager__".  Also, please put screenshots in the PR!

BLOCKERS:  
1.  In accordance with the new team policy, this needs frontend tests -- don't worry about extensive tests, this more about "at least one reasonable one."  You can check out my [admin email PR](https://github.com/broadinstitute/pearl/pull/400) for an example of how to do a basic "confirm if endpoint is mocked, table appears" test
2.  This needs a backend test that we can successfully link studies to kit types to kits to enrollees.  (essentially confirming that all the DAOs and DB constraints match up).  The test flow would be something like:
a.  create kit type
b. create study
c. assign kit type to study
d. list study kit types
e. create enrollee in study
f. create a kit request for them of the given type (these latter two steps might not be necessary since they're essentially being what's already done in testCreatSampleKit, but they will be necessary if we're going to add server-side logic confirm the kit type for an enrollee is allowed by the study)

File: populate/src/main/java/bio/terra/pearl/populate/service/KitTypePopulator.java [hunk 12]
```diff
@@ -0,0 +1,47 @@
+package bio.terra.pearl.populate.service;
+
+import bio.terra.pearl.core.dao.kit.KitTypeDao;
+import bio.terra.pearl.core.model.kit.KitType;
+import bio.terra.pearl.populate.service.contexts.FilePopulateContext;
+import org.springframework.stereotype.Service;
+
+import java.io.IOException;
+import java.util.Optional;
+
```
Comment (devonbush) [inline populate/src/main/java/bio/terra/pearl/populate/service/KitTypePopulator.java:10]: Yup, this looks right to me!  Once you add the dsmStatusJson field, you'll want to Override the `preProcessDto` method to convert that field to the String for dsmStatus

File: populate/src/main/resources/seed/portals/ourhealth/studies/ourheart/enrollees/jsalk.json [hunk 13]
```diff
@@ -64,5 +64,13 @@
...
     ],
     "currentPageNo": 6
-  }]
+  }],
+  "kitRequestDtos": [
+    {
+      "creatingAdminUsername": "breilly@broadinstitute.org",
+      "kitTypeName": "saliva",
+      "statusName": "CREATED",
+      "dsmStatus": "{\n  \"kitId\": \"11111111-2222-3333-4444-555555555555\",\n  \"labelDate\": \"2023-05-24\",\n  \"scanDate\": \"2023-05-24\",\n  \"receiveDate\": \"2023-05-24\",\n  \"scannedByEmail\": \"\",\n  \"receivedByEmail\": \"\",\n  \"labeledByEmail\": \"\",\n  \"deactivationDate\": \"\",\n  \"deactivatedBy\": \"\",\n  \"deactivationReason\": \"\",\n  \"trackingNumbers\": [],\n  \"returnTrackingNumber\": \"\",\n  \"errors\": [\n    {\n      \"errorDate\": \"2023-05-24\",\n      \"errorMessage\": \"\"\n    }\n  ],\n  \"discardDate\": \"2023-05-24\",\n  \"discardReason\": \"\"\n}\n"
```
Comment (devonbush) [inline populate/src/main/resources/seed/portals/ourhealth/studies/ourheart/enrollees/jsalk.json]: To make this more palatable, I'd add a field "dsmStatusJson" to KitRequestPopDto that's a JSONNode.  then the populator can convert that field to a string.  `HtmlSectionPopDto` is another place we do that.  Then you'll be able to specify DSM status as regular json rather than this extended string.

File: populate/src/main/java/bio/terra/pearl/populate/dto/kit/KitRequestDto.java [hunk 14]
```diff
@@ -0,0 +1,13 @@
+package bio.terra.pearl.populate.dto.kit;
+
+import bio.terra.pearl.core.model.kit.KitRequest;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Getter @Setter @NoArgsConstructor
+public class KitRequestDto extends KitRequest {
```
Comment (devonbush) [inline populate/src/main/java/bio/terra/pearl/populate/dto/kit/KitRequestDto.java]: rename this to KitRequestPopDto to clarify that it's a DTO only intended for populate files.  (which is kind of redundant given its package location, but I've found stuff like that makes find-by-name in IDEs much easier.)

**Review by devonbush — COMMENTED** [review]
adding some comments on the populators which I missed in the first review

File: ui-admin/src/study/participants/KitRequests.tsx (hunk 6 above)
Comment (breilly2) [reply ui-admin/src/study/participants/KitRequests.tsx]: I swear I saw somewhere that column definitions should be memoized, but maybe I was looking at the old v7 docs instead of the v8 docs. 🤦‍♂️

File: core/src/main/java/bio/terra/pearl/core/config/DataStudyManagerConfig.java (hunk 8 above)
Comment (breilly2) [reply core/src/main/java/bio/terra/pearl/core/config/DataStudyManagerConfig.java]: It's in config because originally I was trying to use `@ConfigurationProperties`, but that doesn't work when it's coming from `core`. I'll move it into `DataStudyManagerClient`.

File: core/src/main/java/bio/terra/pearl/core/service/kit/DataStudyManagerClient.java (hunk 9 above)
Comment (breilly2) [reply core/src/main/java/bio/terra/pearl/core/service/kit/DataStudyManagerClient.java]: Fair enough. I wasn't sure how much to lean into naming things with "Pepper", and was trying to avoid the uppercase-acronym-inside-camelCase problem.

File: core/src/main/java/bio/terra/pearl/core/service/kit/DataStudyManagerKitRequest.java (hunk 10 above)
Comment (breilly2) [reply core/src/main/java/bio/terra/pearl/core/service/kit/DataStudyManagerKitRequest.java]: I'm trying to serialize two different things: the sent-to address (saved in Juniper) and the web request to DSM. The sent-to address is a subset of the fields in the overall request and is not a separate structure, thus the copying. Alternatives:

* save the entire DSM request in Juniper instead of just the address
* use a bean copying utility library to copy the properties whose names match

File: ui-admin/src/study/participants/KitRequests.tsx (hunk 6 above)
Comment (devonbush) [reply ui-admin/src/study/participants/KitRequests.tsx]: it's correct that you have to memoize them *if* they are within the component function.  But these don't need to be declared within the component function, they can be an external const.

File: populate/src/main/resources/seed/portals/ourhealth/studies/ourheart/enrollees/jsalk.json (hunk 13 above)
Comment (breilly2) [reply populate/src/main/resources/seed/portals/ourhealth/studies/ourheart/enrollees/jsalk.json]: Yes, that is _much_ better. Great suggestion!

File: ui-admin/src/api/api.tsx [hunk 15]
```diff
@@ -435,8 +452,31 @@ export default {
...
-    const url = `${baseStudyEnvUrl(portalShortcode, studyShortcode, envName)
-    }/enrollees/${enrolleeShortcode}/changeRecords`
+    const url =
+      `${baseStudyEnvUrl(portalShortcode, studyShortcode, envName)}/enrollees/${enrolleeShortcode}/changeRecords`
+    const response = await fetch(url, this.getGetInit())
+    return await this.processJsonResponse(response)
+  },
+
+  async createKitRequest(portalShortcode: string, studyShortcode: string, envName: string,
+                         enrolleeShortcode: string, kitType: string): Promise<string> {
```
Comment (breilly2) [inline ui-admin/src/api/api.tsx]: I want to point out that I don't agree that the suggested indentation is better. I had this:
```
  async createKitRequest(portalShortcode: string, studyShortcode: string, envName: string,
                         enrolleeShortcode: string, kitType: string): Promise<string> {
    const params = new URLSearchParams({ kitType })
    const url =
    ...
```
which makes it clear that the parameter list is split across two lines.

The suggested formatting is:
```
  async createKitRequest(portalShortcode: string, studyShortcode: string, envName: string,
    enrolleeShortcode: string, kitType: string): Promise<string> {
    const params = new URLSearchParams({ kitType })
    const url =
    ...
```
which makes the second line of parameters look like the first line of the body.

An alternative would be one parameter per line:
```
  async createKitRequest(
    portalShortcode: string,
    studyShortcode: string,
    envName: string,
    enrolleeShortcode: string,
    kitType: string
  ): Promise<string> {
    const params = new URLSearchParams({ kitType })
    const url =
    ...
```
but it seems excessive to me to take 7 lines to describe a 5 line function (though I'll probably end up going with this style because I think it's the least bad that the linter doesn't complain about).

File: populate/src/main/java/bio/terra/pearl/populate/service/KitTypePopulator.java (hunk 12 above)
Comment (breilly2) [reply populate/src/main/java/bio/terra/pearl/populate/service/KitTypePopulator.java:10]: I ended up not needing a separate `KitRequestPopulator` because it's rolled into `EnrolleePopulator` to avoid duplicating profile/address information (although it is technically possible for a kit request to have a different address than the current profile address). Although now that I think about it, the current enrollee could be passed to `KitRequesntPopulator` through a custom `FilePopulateContext`... but I'm inclined to say that this works well enough for now.

File: core/src/main/java/bio/terra/pearl/core/service/kit/DataStudyManagerKitRequest.java (hunk 10 above)
Comment (breilly2) [reply core/src/main/java/bio/terra/pearl/core/service/kit/DataStudyManagerKitRequest.java]: Devon, do you have any preference for how to resolve this?

**Review by devonbush — APPROVED** [review]
Tests look good, thanks!   One test to keep in mind for later iterations of this kit functionality is to add some assertions to the PopulateOurHealthTest that the kit request got successfully attached to the enrollee.

File: core/src/main/java/bio/terra/pearl/core/service/kit/DataStudyManagerKitRequest.java (hunk 10 above)
Comment (devonbush) [reply core/src/main/java/bio/terra/pearl/core/service/kit/DataStudyManagerKitRequest.java]: Now that I understand what you're doing, this seems fine the way you have it.

File: api-admin/src/main/resources/application.yml [hunk 16]
```diff
@@ -20,6 +20,9 @@ env:
     storageContainerName: ${STORAGE_CONTAINER_NAME:}
     billingProfileId: ${BILLING_PROFILE_ID:550212cf-45dc-478f-9fe6-9b3290e2a1fe}
     deploymentZone: ${DEPLOYMENT_ZONE:}
+  dsm:
+    basePath: ${DSM_ADDRESS:https://dsm-dev.datadonationplatform.org/ddp}
```
Comment (MatthewBemis) [inline api-admin/src/main/resources/application.yml:24]: I've had to think about this for TDR as well: how do we ensure that our prod Juniper instance doesn't send data to dev DSM? It looks like this would silently send data to dev if we forget to explicitly set the dsm address for prod (a potentially easy thing to do)

**Review by MatthewBemis — APPROVED** [review]
Overall looks good, I just have one question

File: api-admin/src/main/resources/application.yml (hunk 16 above)
Comment (breilly2) [reply api-admin/src/main/resources/application.yml:24]: That's a great point. I think it's possible to use `application-development.yml` as an overlay when running locally, but won't be bundled in any artifact that is run in any other environment. I ran a quick experiment and it didn't work the way I wanted it to. Could research this some more, but I'm not going to spend that time right now.

Comment (breilly2) [reply api-admin/src/main/resources/application.yml:24]: That's a great point. I think it's possible to use `application-development.yml` as an overlay when running locally, but won't be bundled in any artifact that is run in any other environment. I ran a quick experiment and it didn't work the way I wanted it to. Could research this some more, but I'm not going to spend that time right now.
